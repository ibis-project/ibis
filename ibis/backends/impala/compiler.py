from __future__ import annotations

import ibis.expr.operations as ops
from ibis.backends.base.sql.compiler import Compiler, ExprTranslator, TableSetFormatter
from ibis.backends.base.sql.registry import binary_infix_ops, operation_registry, unary


class ImpalaTableSetFormatter(TableSetFormatter):
    def _get_join_type(self, op):
        jname = self._join_names[type(op)]

        # Impala requires this
        if not op.predicates:
            jname = self._join_names[ops.CrossJoin]

        return jname

    def _format_in_memory_table(self, op):
        if op.data:
            return super()._format_in_memory_table(op)

        schema = op.schema
        names = schema.names
        types = schema.types
        rows = [
            f"{self._translate(ops.Cast(ops.Literal(None, dtype=dtype), to=dtype))} AS {name}"
            for name, dtype in zip(map(self._quote_identifier, names), types)
        ]
        return f"(SELECT * FROM (SELECT {', '.join(rows)}) AS _ LIMIT 0)"


class ImpalaExprTranslator(ExprTranslator):
    _registry = {**operation_registry, **binary_infix_ops, ops.Hash: unary("fnv_hash")}
    _forbids_frame_clause = (
        *ExprTranslator._forbids_frame_clause,
        ops.Lag,
        ops.Lead,
        ops.FirstValue,
        ops.LastValue,
    )
    _unsupported_reductions = (
        ops.ApproxMedian,
        ops.ApproxCountDistinct,
        ops.GroupConcat,
    )
    _dialect_name = "hive"
    _quote_identifiers = True


rewrites = ImpalaExprTranslator.rewrites


@rewrites(ops.FloorDivide)
def _floor_divide(op):
    return ops.Floor(ops.Divide(op.left, op.right))


class ImpalaCompiler(Compiler):
    translator_class = ImpalaExprTranslator
    table_set_formatter_class = ImpalaTableSetFormatter
