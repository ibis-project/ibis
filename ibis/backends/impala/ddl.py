from __future__ import annotations

# Copyright 2014 Cloudera Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import json

from ibis.backends.base.sql.ddl import (
    AlterTable,
    BaseDDL,
    CreateTable,
    CreateTableWithSchema,
    DropFunction,
    format_partition,
    format_schema,
    format_tblproperties,
)
from ibis.backends.base.sql.registry import type_to_sql_string


class CreateTableParquet(CreateTable):
    def __init__(
        self,
        table_name,
        path,
        example_file=None,
        example_table=None,
        schema=None,
        external=True,
        **kwargs,
    ):
        super().__init__(
            table_name,
            external=external,
            format="parquet",
            path=path,
            **kwargs,
        )
        self.example_file = example_file
        self.example_table = example_table
        self.schema = schema

    @property
    def _pieces(self):
        if self.example_file is not None:
            yield f"LIKE PARQUET '{self.example_file}'"
        elif self.example_table is not None:
            yield f"LIKE {self.example_table}"
        elif self.schema is not None:
            yield format_schema(self.schema)
        else:
            raise NotImplementedError

        yield self._storage()
        yield self._location()


class DelimitedFormat:
    def __init__(
        self,
        path,
        delimiter=None,
        escapechar=None,
        na_rep=None,
        lineterminator=None,
    ):
        self.path = path
        self.delimiter = delimiter
        self.escapechar = escapechar
        self.lineterminator = lineterminator
        self.na_rep = na_rep

    def to_ddl(self):
        yield "ROW FORMAT DELIMITED"

        if self.delimiter is not None:
            yield f"FIELDS TERMINATED BY '{self.delimiter}'"

        if self.escapechar is not None:
            yield f"ESCAPED BY '{self.escapechar}'"

        if self.lineterminator is not None:
            yield f"LINES TERMINATED BY '{self.lineterminator}'"

        yield "STORED AS TEXTFILE"
        yield f"LOCATION '{self.path}'"

        if self.na_rep is not None:
            props = {"serialization.null.format": self.na_rep}
            yield format_tblproperties(props)


class AvroFormat:
    def __init__(self, path, avro_schema):
        self.path = path
        self.avro_schema = avro_schema

    def to_ddl(self):
        yield "STORED AS AVRO"
        yield f"LOCATION '{self.path}'"

        schema = json.dumps(self.avro_schema, indent=2, sort_keys=True)
        schema = "\n".join(x.rstrip() for x in schema.splitlines())

        props = {"avro.schema.literal": schema}
        yield format_tblproperties(props)


class ParquetFormat:
    def __init__(self, path):
        self.path = path

    def to_ddl(self):
        yield "STORED AS PARQUET"
        yield f"LOCATION '{self.path}'"


class CreateTableDelimited(CreateTableWithSchema):
    def __init__(
        self,
        table_name,
        path,
        schema,
        delimiter=None,
        escapechar=None,
        lineterminator=None,
        na_rep=None,
        external=True,
        **kwargs,
    ):
        table_format = DelimitedFormat(
            path,
            delimiter=delimiter,
            escapechar=escapechar,
            lineterminator=lineterminator,
            na_rep=na_rep,
        )
        super().__init__(table_name, schema, table_format, external=external, **kwargs)


class CreateTableAvro(CreateTable):
    def __init__(self, table_name, path, avro_schema, external=True, **kwargs):
        super().__init__(table_name, external=external, **kwargs)
        self.table_format = AvroFormat(path, avro_schema)

    @property
    def _pieces(self):
        yield "\n".join(self.table_format.to_ddl())


class LoadData(BaseDDL):
    """Generate DDL for LOAD DATA command.

    Cannot be cancelled
    """

    def __init__(
        self,
        table_name,
        path,
        database=None,
        partition=None,
        partition_schema=None,
        overwrite=False,
    ):
        self.table_name = table_name
        self.database = database
        self.path = path

        self.partition = partition
        self.partition_schema = partition_schema

        self.overwrite = overwrite

    def compile(self):
        overwrite = "OVERWRITE " if self.overwrite else ""

        if self.partition is not None:
            partition = "\n" + format_partition(self.partition, self.partition_schema)
        else:
            partition = ""

        scoped_name = self._get_scoped_name(self.table_name, self.database)
        return "LOAD DATA INPATH '{}' {}INTO TABLE {}{}".format(
            self.path, overwrite, scoped_name, partition
        )


class PartitionProperties(AlterTable):
    def __init__(
        self,
        table,
        partition,
        partition_schema,
        location=None,
        format=None,
        tbl_properties=None,
        serde_properties=None,
    ):
        super().__init__(
            table,
            location=location,
            format=format,
            tbl_properties=tbl_properties,
            serde_properties=serde_properties,
        )
        self.partition = partition
        self.partition_schema = partition_schema

    def _compile(self, cmd, property_prefix=""):
        part = format_partition(self.partition, self.partition_schema)
        if cmd:
            part = f"{cmd} {part}"

        props = self._format_properties(property_prefix)
        action = f"{self.table} {part}{props}"
        return self._wrap_command(action)


class AddPartition(PartitionProperties):
    def __init__(self, table, partition, partition_schema, location=None):
        super().__init__(table, partition, partition_schema, location=location)

    def compile(self):
        return self._compile("ADD")


class AlterPartition(PartitionProperties):
    def compile(self):
        return self._compile("", "SET ")


class DropPartition(PartitionProperties):
    def __init__(self, table, partition, partition_schema):
        super().__init__(table, partition, partition_schema)

    def compile(self):
        return self._compile("DROP")


class CacheTable(BaseDDL):
    def __init__(self, table_name, database=None, pool="default"):
        self.table_name = table_name
        self.database = database
        self.pool = pool

    def compile(self):
        scoped_name = self._get_scoped_name(self.table_name, self.database)
        return f"ALTER TABLE {scoped_name} SET CACHED IN '{self.pool}'"


class CreateFunction(BaseDDL):
    _object_type = "FUNCTION"

    def __init__(self, func, name=None, database=None):
        self.func = func
        self.name = name or func.name
        self.database = database

    def _impala_signature(self):
        scoped_name = self._get_scoped_name(self.name, self.database)
        input_sig = _impala_input_signature(self.func.inputs)
        output_sig = type_to_sql_string(self.func.output)

        return f"{scoped_name}({input_sig}) returns {output_sig}"


class CreateUDF(CreateFunction):
    def compile(self):
        create_decl = "CREATE FUNCTION"
        impala_sig = self._impala_signature()
        param_line = f"location '{self.func.lib_path}' symbol='{self.func.so_symbol}'"
        return f"{create_decl} {impala_sig} {param_line}"


class CreateUDA(CreateFunction):
    def compile(self):
        create_decl = "CREATE AGGREGATE FUNCTION"
        impala_sig = self._impala_signature()
        tokens = [f"location '{self.func.lib_path}'"]

        fn_names = (
            "init_fn",
            "update_fn",
            "merge_fn",
            "serialize_fn",
            "finalize_fn",
        )

        for fn in fn_names:
            value = getattr(self.func, fn)
            if value is not None:
                tokens.append(f"{fn}='{value}'")

        joined_tokens = "\n".join(tokens)
        return f"{create_decl} {impala_sig} {joined_tokens}"


class DropFunction(DropFunction):
    def _impala_signature(self):
        full_name = self._get_scoped_name(self.name, self.database)
        input_sig = _impala_input_signature(self.inputs)
        return f"{full_name}({input_sig})"


class ListFunction(BaseDDL):
    def __init__(self, database, like=None, aggregate=False):
        self.database = database
        self.like = like
        self.aggregate = aggregate

    def compile(self):
        statement = "SHOW "
        if self.aggregate:
            statement += "AGGREGATE "
        statement += f"FUNCTIONS IN {self.database}"
        if self.like:
            statement += f" LIKE '{self.like}'"
        return statement


def _impala_input_signature(inputs):
    # TODO: varargs '{}...'.format(val)
    return ", ".join(map(type_to_sql_string, inputs))
