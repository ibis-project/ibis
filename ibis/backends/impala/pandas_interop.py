from __future__ import annotations

# Copyright 2014 Cloudera Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import contextlib
import csv
import os
import tempfile
from posixpath import join as pjoin

import ibis.common.exceptions as com
import ibis.expr.schema as sch
from ibis import util
from ibis.config import options


class DataFrameWriter:
    """Interface class for writing pandas objects to Impala tables.

    Notes
    -----
    Class takes ownership of any temporary data written to HDFS
    """

    def __init__(self, client, df):
        self.client = client
        self.df = df
        self.temp_hdfs_dirs = set()

    def write_temp_csv(self):
        temp_hdfs_dir = pjoin(options.impala.temp_hdfs_path, f"pandas_{util.guid()}")
        self.client.hdfs.mkdir(temp_hdfs_dir)

        # Keep track of the temporary HDFS file
        self.temp_hdfs_dirs.add(temp_hdfs_dir)

        # Write the file to HDFS
        hdfs_path = pjoin(temp_hdfs_dir, "0.csv")

        self.write_csv(hdfs_path)

        return temp_hdfs_dir

    def write_csv(self, path):
        # Use a temporary dir instead of a temporary file
        # to provide Windows support and avoid #2267
        # https://github.com/ibis-project/ibis/issues/2267
        with tempfile.TemporaryDirectory() as f:
            # Write the DataFrame to the temporary file path
            tmp_file_path = os.path.join(f, "impala_temp_file.csv")
            if options.verbose:
                util.log(f"Writing DataFrame to temporary directory {tmp_file_path}")

            self.df.to_csv(
                tmp_file_path,
                header=False,
                index=False,
                sep=",",
                quoting=csv.QUOTE_NONE,
                escapechar="\\",
                na_rep="#NULL",
            )

            if options.verbose:
                util.log(f"Writing CSV to: {path}")

            self.client.hdfs.put(tmp_file_path, path)
        return path

    def get_schema(self):
        # define a temporary table using delimited data
        return sch.infer(self.df)

    def delimited_table(self, csv_dir, database=None):
        return self.client.delimited_file(
            csv_dir,
            self.get_schema(),
            name=f"ibis_tmp_pandas_{util.guid()}",
            database=database,
            delimiter=",",
            na_rep="#NULL",
            escapechar="\\\\",
            external=True,
            persist=False,
        )

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        with contextlib.suppress(com.IbisError):
            self.cleanup()
        return False

    def cleanup(self):
        while self.temp_hdfs_dirs:
            self.client.hdfs.rm(self.temp_hdfs_dirs.pop(), recursive=True)
