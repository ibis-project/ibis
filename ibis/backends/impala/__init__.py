"""Impala backend"""
import ibis.config
from ibis.backends.base import BaseBackend

# these objects are exposed in the public API and are not used in the module
from .client import (  # noqa: F401
    ImpalaClient,
    ImpalaConnection,
    ImpalaDatabase,
    ImpalaDatabaseTable,
    ImpalaTable,
)
from .hdfs import HDFS, WebHDFS, hdfs_connect
from .udf import *  # noqa: F401,F403


class Backend(BaseBackend):
    name = 'impala'
    kind = 'sql'
    client = ImpalaClient
    database_class = ImpalaDatabase
    table_class = ImpalaDatabaseTable
    table_expr_class = ImpalaTable
    HDFS = HDFS
    WebHDFS = WebHDFS

    def hdfs_connect(self, *args, **kwargs):
        return hdfs_connect(*args, **kwargs)

    def connect(
        self,
        host='localhost',
        port=21050,
        database='default',
        timeout=45,
        use_ssl=False,
        ca_cert=None,
        user=None,
        password=None,
        auth_mechanism='NOSASL',
        kerberos_service_name='impala',
        pool_size=8,
        hdfs_client=None,
    ):
        """Create an ImpalaClient for use with Ibis.

        Parameters
        ----------
        host : str, optional
            Host name of the impalad or HiveServer2 in Hive
        port : int, optional
            Impala's HiveServer2 port
        database : str, optional
            Default database when obtaining new cursors
        timeout : int, optional
            Connection timeout in seconds when communicating with HiveServer2
        use_ssl : bool, optional
            Use SSL when connecting to HiveServer2
        ca_cert : str, optional
            Local path to 3rd party CA certificate or copy of server
            certificate for self-signed certificates. If SSL is enabled, but
            this argument is ``None``, then certificate validation is skipped.
        user : str, optional
            LDAP user to authenticate
        password : str, optional
            LDAP password to authenticate
        auth_mechanism : str, optional
            {'NOSASL' <- default, 'PLAIN', 'GSSAPI', 'LDAP'}.
            Use NOSASL for non-secured Impala connections.  Use PLAIN for
            non-secured Hive clusters.  Use LDAP for LDAP authenticated
            connections.  Use GSSAPI for Kerberos-secured clusters.
        kerberos_service_name : str, optional
            Specify particular impalad service principal.

        Examples
        --------
        >>> import ibis
        >>> import os
        >>> hdfs_host = os.environ.get('IBIS_TEST_NN_HOST', 'localhost')
        >>> hdfs_port = int(os.environ.get('IBIS_TEST_NN_PORT', 50070))
        >>> impala_host = os.environ.get('IBIS_TEST_IMPALA_HOST', 'localhost')
        >>> impala_port = int(os.environ.get('IBIS_TEST_IMPALA_PORT', 21050))
        >>> hdfs = ibis.impala.hdfs_connect(host=hdfs_host, port=hdfs_port)
        >>> hdfs  # doctest: +ELLIPSIS
        <ibis.filesystems.WebHDFS object at 0x...>
        >>> client = ibis.impala.connect(
        ...     host=impala_host,
        ...     port=impala_port,
        ...     hdfs_client=hdfs,
        ... )
        >>> client  # doctest: +ELLIPSIS
        <ibis.impala.client.ImpalaClient object at 0x...>

        Returns
        -------
        ImpalaClient
        """
        params = {
            'host': host,
            'port': port,
            'database': database,
            'timeout': timeout,
            'use_ssl': use_ssl,
            'ca_cert': ca_cert,
            'user': user,
            'password': password,
            'auth_mechanism': auth_mechanism,
            'kerberos_service_name': kerberos_service_name,
        }

        con = ImpalaConnection(pool_size=pool_size, **params)
        try:
            client = ImpalaClient(
                backend=self, con=con, hdfs_client=hdfs_client
            )
        except Exception:
            con.close()
            raise
        return client

    def register_options(self):
        ibis.config.register_option(
            'temp_db',
            '__ibis_tmp',
            'Database to use for temporary tables, views. functions, etc.',
        )
        ibis.config.register_option(
            'temp_hdfs_path',
            '/tmp/ibis',
            'HDFS path for storage of temporary data',
        )
