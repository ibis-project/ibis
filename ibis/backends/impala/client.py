from __future__ import annotations

import contextlib
import traceback
from typing import TYPE_CHECKING

import sqlalchemy as sa
import sqlglot as sg

import ibis.common.exceptions as com
import ibis.expr.schema as sch
import ibis.expr.types as ir
from ibis import util
from ibis.backends.base import Database
from ibis.backends.base.sql.compiler import DDL, DML
from ibis.backends.base.sql.ddl import AlterTable, InsertSelect
from ibis.backends.impala import ddl
from ibis.backends.impala.compat import HS2Error, impyla

if TYPE_CHECKING:
    import pandas as pd


class ImpalaDatabase(Database):
    def create_table(self, name: str, obj=None, **kwargs) -> ir.Table:
        """Dispatch to ImpalaClient.create_table.

        See that function's docstring for more
        """
        return self.client.create_table(name, obj=obj, database=self.name, **kwargs)

    def list_udfs(self, like=None):
        return self.client.list_udfs(like=like, database=self.name)

    def list_udas(self, like=None):
        return self.client.list_udas(like=like, database=self.name)


class ImpalaConnection:
    """Database connection wrapper."""

    def __init__(self, pool_size=8, database="default", **params):
        self.params = params
        self.database = database
        self.options = {}
        self.pool = sa.pool.QueuePool(
            self._new_cursor,
            pool_size=pool_size,
            # disable invoking rollback, because any transactions in impala are
            # automatic:
            # https://impala.apache.org/docs/build/html/topics/impala_transactions.html
            reset_on_return=False,
        )

        @sa.event.listens_for(self.pool, "checkout")
        def _(dbapi_connection, *_):
            """Update `dbapi_connection` options if they don't match `self`."""
            if (options := self.options) != dbapi_connection.options:
                dbapi_connection.options = options.copy()
                dbapi_connection.set_options()

    def set_options(self, options):
        self.options.update(options)

    def close(self):
        """Close all idle Impyla connections."""
        self.pool.dispose()

    def disable_codegen(self, disabled=True):
        self.options["DISABLE_CODEGEN"] = str(int(disabled))

    def execute(self, query):
        if isinstance(query, (DDL, DML)):
            query = query.compile()

        util.log(query)

        cursor = self.pool.connect()
        try:
            cursor.execute(query)
        except Exception:
            cursor.close()
            util.log(f"Exception caused by {query}: {traceback.format_exc()}")
            raise

        return cursor

    def fetchall(self, query):
        cur = self.execute(query)
        try:
            results = cur.fetchall()
        finally:
            cur.close()
        return results

    def _new_cursor(self):
        params = self.params.copy()
        con = impyla.connect(database=self.database, **params)

        # make sure the connection works
        cursor = con.cursor(user=params.get("user"), convert_types=True)
        cursor.ping()

        wrapper = ImpalaCursor(cursor, con, self.database, self.options.copy())
        wrapper.set_options()
        return wrapper

    def release(self, cur):  # pragma: no cover
        pass


class ImpalaCursor:
    def __init__(self, cursor, impyla_con, database, options):
        self._cursor = cursor
        self.impyla_con = impyla_con
        self.database = database
        self.options = options

    def __del__(self):
        with contextlib.suppress(Exception):
            self.close()

    def close(self):
        try:
            self._cursor.close()
        except HS2Error as e:
            # connection was closed elsewhere
            already_closed_messages = [
                "invalid query handle",
                "invalid session",
            ]
            for message in already_closed_messages:
                if message in e.args[0].lower():
                    break
            else:
                raise

    def set_options(self):
        for k, v in self.options.items():
            query = f"SET {k} = {v!r}"
            self._cursor.execute(query)

    @property
    def description(self):
        return self._cursor.description

    def release(self):
        pass

    def execute(self, stmt):
        self._cursor.execute_async(stmt)
        self._wait_synchronous()

    def _wait_synchronous(self):
        # Wait to finish, but cancel if KeyboardInterrupt
        try:
            self._cursor._wait_to_finish()
        except KeyboardInterrupt:
            util.log("Canceling query")
            self.cancel()
            raise

    def is_finished(self):
        return not self.is_executing()

    def is_executing(self):
        return self._cursor.is_executing()

    def cancel(self):
        self._cursor.cancel_operation()

    def fetchone(self):
        return self._cursor.fetchone()

    def fetchall(self, columnar=False):
        if columnar:
            return self._cursor.fetchcolumnar()
        else:
            return self._cursor.fetchall()


class ImpalaTable(ir.Table):
    """A physical table in the Impala-Hive metastore."""

    @property
    def _qualified_name(self) -> str:
        op = self.op()
        return sg.table(op.name, db=op.namespace).sql(dialect="hive")

    @property
    def _unqualified_name(self) -> str:
        return self.op().name

    @property
    def _client(self):
        return self.op().source

    @property
    def _database(self) -> str:
        return self.op().namespace

    def compute_stats(self, incremental=False):
        """Invoke Impala COMPUTE STATS command on the table."""
        return self._client.compute_stats(self._qualified_name, incremental=incremental)

    def invalidate_metadata(self):
        self._client.invalidate_metadata(self._qualified_name)

    def refresh(self):
        self._client.refresh(self._qualified_name)

    def metadata(self):
        """Return results of `DESCRIBE FORMATTED` statement."""
        return self._client.describe_formatted(self._qualified_name)

    describe_formatted = metadata

    def files(self):
        """Return results of SHOW FILES statement."""
        return self._client.show_files(self._qualified_name)

    def drop(self):
        """Drop the table from the database."""
        self._client.drop_table_or_view(self._qualified_name)

    def truncate(self):
        self._client.truncate_table(self._qualified_name)

    def insert(
        self,
        obj=None,
        overwrite=False,
        partition=None,
        values=None,
        validate=True,
    ):
        """Insert into an Impala table.

        Parameters
        ----------
        obj
            Table expression or DataFrame
        overwrite
            If True, will replace existing contents of table
        partition
            For partitioned tables, indicate the partition that's being
            inserted into, either with an ordered list of partition keys or a
            dict of partition field name to value. For example for the
            partition (year=2007, month=7), this can be either (2007, 7) or
            {'year': 2007, 'month': 7}.
        values
            Unsupported and unused
        validate
            If True, do more rigorous validation that schema of table being
            inserted is compatible with the existing table

        Examples
        --------
        Append to an existing table

        >>> t.insert(table_expr)  # quartodoc: +SKIP # doctest: +SKIP

        Completely overwrite contents

        >>> t.insert(table_expr, overwrite=True)  # quartodoc: +SKIP # doctest: +SKIP
        """
        if values is not None:
            raise NotImplementedError

        with self._client._setup_insert(obj) as expr:
            if validate:
                existing_schema = self.schema()
                insert_schema = expr.schema()
                if not insert_schema.equals(existing_schema):
                    _validate_compatible(insert_schema, existing_schema)

            if partition is not None:
                partition_schema = self.partition_schema()
                partition_schema_names = frozenset(partition_schema.names)
                expr = expr.select(
                    [
                        column
                        for column in expr.columns
                        if column not in partition_schema_names
                    ]
                )
            else:
                partition_schema = None

            ast = self._client.compiler.to_ast(expr)
            select = ast.queries[0]
            statement = InsertSelect(
                self._qualified_name,
                select,
                partition=partition,
                partition_schema=partition_schema,
                overwrite=overwrite,
            )
            self._client._safe_exec_sql(statement.compile())
            return self

    def load_data(self, path, overwrite=False, partition=None):
        """Load data into an Impala table.

        Parameters
        ----------
        path
            Data to load
        overwrite
            Overwrite the existing data in the entire table or indicated
            partition
        partition
            If specified, the partition must already exist
        """
        if partition is not None:
            partition_schema = self.partition_schema()
        else:
            partition_schema = None

        stmt = ddl.LoadData(
            self._qualified_name,
            path,
            partition=partition,
            partition_schema=partition_schema,
            overwrite=overwrite,
        )

        self._client._safe_exec_sql(stmt.compile())
        return self

    @property
    def name(self) -> str:
        return self.op().name

    @property
    def is_partitioned(self):
        """True if the table is partitioned."""
        return self.metadata().is_partitioned

    def partition_schema(self):
        """Return the schema for the partition columns."""
        schema = self.schema()
        result = self.partitions()

        partition_fields = []
        for col in result.columns:
            if col not in schema:
                break
            partition_fields.append((col, schema[col]))

        return sch.Schema(dict(partition_fields))

    def add_partition(self, spec, location=None):
        """Add a new table partition.

        This API creates any necessary new directories in HDFS.

        Partition parameters can be set in a single DDL statement or you can
        use `alter_partition` to set them after the fact.
        """
        part_schema = self.partition_schema()
        stmt = ddl.AddPartition(
            self._qualified_name, spec, part_schema, location=location
        )
        self._client._safe_exec_sql(stmt)
        return self

    def alter(
        self,
        location=None,
        format=None,
        tbl_properties=None,
        serde_properties=None,
    ):
        """Change settings and parameters of the table.

        Parameters
        ----------
        location
            For partitioned tables, you may want the alter_partition function
        format
            Table format
        tbl_properties
            Table properties
        serde_properties
            Serialization/deserialization properties
        """

        def _run_ddl(**kwds):
            stmt = AlterTable(self._qualified_name, **kwds)
            self._client._safe_exec_sql(stmt)
            return self

        return self._alter_table_helper(
            _run_ddl,
            location=location,
            format=format,
            tbl_properties=tbl_properties,
            serde_properties=serde_properties,
        )

    def set_external(self, is_external=True):
        """Toggle the `EXTERNAL` table property."""
        self.alter(tbl_properties={"EXTERNAL": is_external})

    def alter_partition(
        self,
        spec,
        location=None,
        format=None,
        tbl_properties=None,
        serde_properties=None,
    ):
        """Change settings and parameters of an existing partition.

        Parameters
        ----------
        spec
            The partition keys for the partition being modified
        location
            Location of the partition
        format
            Table format
        tbl_properties
            Table properties
        serde_properties
            Serialization/deserialization properties
        """
        part_schema = self.partition_schema()

        def _run_ddl(**kwds):
            stmt = ddl.AlterPartition(self._qualified_name, spec, part_schema, **kwds)
            self._client._safe_exec_sql(stmt)
            return self

        return self._alter_table_helper(
            _run_ddl,
            location=location,
            format=format,
            tbl_properties=tbl_properties,
            serde_properties=serde_properties,
        )

    def _alter_table_helper(self, f, **alterations):
        results = []
        for k, v in alterations.items():
            if v is None:
                continue
            result = f(**{k: v})
            results.append(result)
        return results

    def drop_partition(self, spec):
        """Drop an existing table partition."""
        part_schema = self.partition_schema()
        stmt = ddl.DropPartition(self._qualified_name, spec, part_schema)
        self._client._safe_exec_sql(stmt)
        return self

    def partitions(self):
        """Return information about the table's partitions.

        Raises an exception if the table is not partitioned.
        """
        return self._client.list_partitions(self._qualified_name)

    def stats(self) -> pd.DataFrame:
        """Return results of `SHOW TABLE STATS`.

        If not partitioned, contains only one row.

        Returns
        -------
        DataFrame
            Table statistics
        """
        return self._client.table_stats(self._qualified_name)

    def column_stats(self) -> pd.DataFrame:
        """Return results of `SHOW COLUMN STATS`.

        Returns
        -------
        DataFrame
            Column statistics
        """
        return self._client.column_stats(self._qualified_name)


# ----------------------------------------------------------------------
# ORM-ish usability layer


class ScalarFunction:
    def drop(self):
        pass


class AggregateFunction:
    def drop(self):
        pass


def _validate_compatible(from_schema, to_schema):
    if set(from_schema.names) != set(to_schema.names):
        raise com.IbisInputError("Schemas have different names")

    for name in from_schema:
        lt = from_schema[name]
        rt = to_schema[name]
        if not lt.castable(rt):
            raise com.IbisInputError(f"Cannot safely cast {lt!r} to {rt!r}")
