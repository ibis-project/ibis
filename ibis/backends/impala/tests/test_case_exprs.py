from __future__ import annotations

import pytest

import ibis
import ibis.expr.types as ir
from ibis.backends.impala.compiler import ImpalaCompiler
from ibis.backends.impala.tests.conftest import translate


@pytest.fixture(scope="module")
def table(mockcon):
    return mockcon.table("alltypes")


@pytest.fixture
def simple_case(table):
    return table.g.case().when("foo", "bar").when("baz", "qux").else_("default").end()


@pytest.fixture
def search_case(table):
    t = table
    return ibis.case().when(t.f > 0, t.d * 2).when(t.c < 0, t.a * 2).end()


@pytest.fixture
def tpch_lineitem(mockcon):
    return mockcon.table("tpch_lineitem")


def test_isnull_1_0(table, snapshot):
    expr = table.g.isnull().ifelse(1, 0)

    result = translate(expr)
    snapshot.assert_match(result, "out1.sql")

    # inside some other function
    result = translate(expr.sum())
    snapshot.assert_match(result, "out2.sql")


def test_simple_case(simple_case, snapshot):
    expr = simple_case
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


def test_search_case(search_case, snapshot):
    expr = search_case
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


def test_ifelse_use_if(table, snapshot):
    expr = ibis.ifelse(table.f > 0, table.e, table.a)
    assert isinstance(expr, ir.FloatingValue)

    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "expr_fn",
    [
        pytest.param(lambda f: f.nullif(f), id="nullif_input"),
        pytest.param(lambda f: (f == 0).nullif(f == 0), id="nullif_boolean"),
        pytest.param(lambda f: (f != 0).nullif(f == 0), id="nullif_negate_boolean"),
    ],
)
def test_nullif_ifnull(tpch_lineitem, expr_fn, snapshot):
    expr = expr_fn(tpch_lineitem.l_quantity)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


@pytest.mark.parametrize(
    "expr_fn",
    [
        pytest.param(lambda t: t.l_quantity.fillna(0), id="fillna_l_quantity"),
        pytest.param(
            lambda t: t.l_extendedprice.fillna(0), id="fillna_l_extendedprice"
        ),
        pytest.param(
            lambda t: t.l_extendedprice.fillna(0.0), id="fillna_l_extendedprice_double"
        ),
    ],
)
def test_decimal_fillna_cast_arg(tpch_lineitem, expr_fn, snapshot):
    expr = expr_fn(tpch_lineitem)
    result = translate(expr)
    snapshot.assert_match(result, "out.sql")


def test_identical_to(mockcon, snapshot):
    t = mockcon.table("functional_alltypes")
    expr = t.tinyint_col.identical_to(t.double_col).name("tmp")
    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")


def test_identical_to_special_case(snapshot):
    expr = ibis.NA.cast("int64").identical_to(ibis.NA.cast("int64")).name("tmp")
    result = ImpalaCompiler.to_sql(expr)
    snapshot.assert_match(result, "out.sql")
