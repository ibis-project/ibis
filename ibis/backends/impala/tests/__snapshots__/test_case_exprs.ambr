# name: test_decimal_fillna_cast_arg[fillna_l_extendedprice]
  'isnull(`l_extendedprice`, 0)'
# ---
# name: test_decimal_fillna_cast_arg[fillna_l_extendedprice_double]
  'isnull(`l_extendedprice`, 0.0)'
# ---
# name: test_decimal_fillna_cast_arg[fillna_l_quantity]
  'isnull(`l_quantity`, 0)'
# ---
# name: test_identical_to
  '''
  SELECT `tinyint_col` IS NOT DISTINCT FROM `double_col` AS `tmp`
  FROM functional_alltypes
  '''
# ---
# name: test_identical_to_special_case
  'SELECT TRUE AS `tmp`'
# ---
# name: test_isnull_1_0
  'if(`g` IS NULL, 1, 0)'
# ---
# name: test_isnull_1_0.1
  'sum(if(`g` IS NULL, 1, 0))'
# ---
# name: test_nullif_ifnull[nullif_boolean]
  'nullif(`l_quantity` = 0, `l_quantity` = 0)'
# ---
# name: test_nullif_ifnull[nullif_input]
  'nullif(`l_quantity`, `l_quantity`)'
# ---
# name: test_nullif_ifnull[nullif_negate_boolean]
  'nullif(`l_quantity` != 0, `l_quantity` = 0)'
# ---
# name: test_search_case
  '''
  CASE
    WHEN `f` > 0 THEN `d` * 2
    WHEN `c` < 0 THEN `a` * 2
    ELSE CAST(NULL AS bigint)
  END
  '''
# ---
# name: test_simple_case
  '''
  CASE `g`
    WHEN 'foo' THEN 'bar'
    WHEN 'baz' THEN 'qux'
    ELSE 'default'
  END
  '''
# ---
# name: test_where_use_if
  'if(`f` > 0, `e`, `a`)'
# ---
